plugins/mongodb/mongodbembed.cpp

```diff
@@ -0,0 +1,2302 @@
+    {
+        ECLPluginDefinitionBlockEx * pbx = (ECLPluginDefinitionBlockEx *) pb;
+        pbx->compatibleVersions = COMPATIBLE_VERSIONS;
+    }
+    else if (pb->size != sizeof(ECLPluginDefinitionBlock))
+        return false;
+
+    pb->magicVersion = PLUGIN_VERSION;
+    pb->version = VERSION;
+    pb->moduleName = MODULE_NAME;
+    pb->ECL = nullptr;
+    pb->flags = PLUGIN_IMPLICIT_MODULE;
+    pb->description = MODULE_DESCRIPTION;
+    return true;
+}
+
+// Place example code declarations here for doxygen. Source files should be placed in /mongodb/examples.
+/** \example mongodb-test.ecl
+ * Here is some example code of the plugin being used.
+ */
+namespace mongodbembed
+{
+    const time_t OBJECT_EXPIRE_TIMEOUT_SECONDS = 60 * 2; // Two minutes
+    static std::once_flag CONNECTION_CACHE_INIT_FLAG;
+
+    //--------------------------------------------------------------------------
+    // Plugin Classes
+    //--------------------------------------------------------------------------
+
+    /**
+     * @brief Takes an exception object and outputs the error to the user
+     * 
+     * @param e mongocxx::exception has some values that are useful to look for
+     */
+    void reportIfQueryFailure(const mongocxx::exception &e)
+    {
+        if (e.code() == mongocxx::error_code::k_invalid_collection_object) 
+        {
+            failx("Collection not found: %s",e.what());
+        }
+        if (e.code().value() == 26) 
+        {
+            failx("NamespaceNotFound: %s",e.what());
+        }
+        if (e.code().value() == 11000) 
+        {
+            failx("Duplicate Key: %s",e.what());
+        }
+        failx("Exception: %s", e.what());
+    }
+
+    /**
+     * @brief deserialize MongoDB Extended JSON relaxed format into normal JSON for the RowStream class
+     * Examples of EJSON relaxed format are:
+     * "dateField": {"$date":{"$numberLong":"1565546054692"}}
+     * "doubleField": {"$numberDouble":"10.5"}
+     * For more documentation on EJSON. https://www.mongodb.com/docs/manual/reference/mongodb-extended-json/
+     * 
+     * @param result Reference to a result string where the standard JSON should be written.
+     * @param row Pointer to the beginning of the result row.
+     */
+    void deserializeEJSON(std::string &result, const char * row)
+    {
+        auto start = row;
+        int depth = 0; // Keeps track of depth so we don't exit on wrong '}'
+        const char * lastBrkt;
+
+        while (*start != '}' || depth > 1) 
+        {
+            if (*start == '{') // If we see a document increase the depth
+            {
+                depth++;
+                lastBrkt = start; // Keep track of last bracket in case we need to backtrack
+            }
+            if (*start == '$' && *(start-1) == '\"') // Look for '$' to mark the start of a datatype
+            { 
+                auto end = start;
+                while (*end != '\"') end++;
```

Comment:
Will this read off the end of the string if no " is found?

Reply:
Technically yes, if no " is found it reads off the end, but because we are only reading MongoDB EJSON objects the datatype names are always surrounded by ". An example of the objects this function processes is:
`{"$date":{"$numberLong":"-1577923200000"}}`